src/main/sensors/gyroanalyse.c

```diff
@@ -1,35 +1,320 @@
+static gyroFftData_t fftResult[3];
+static uint16_t fftMaxFreq = 0;             // nyquist rate
+static uint16_t fftIdx = 0;                 // use a circular buffer for the last FFT_WINDOW_SIZE samples
+
+
+// accumulator for oversampled data => no aliasing and less noise
+static float fftAcc[3] = {0, 0, 0};
+static int fftAccCount = 0;
+static int fftSamplingScale;
+
+// bandpass filter gyro data
+static biquadFilter_t fftGyroFilter[3];
+
+// filter for smoothing frequency estimation
+static biquadFilter_t fftFreqFilter[3];
+
+// Hanning window, see https://en.wikipedia.org/wiki/Window_function#Hann_.28Hanning.29_window
+static float hanningWindow[FFT_WINDOW_SIZE];
+
+void initHanning() {
+    for (int i = 0; i < FFT_WINDOW_SIZE; ++i) {
+        hanningWindow[i] = (0.5 - 0.5 * cosf(2 * M_PIf * i / (FFT_WINDOW_SIZE - 1)));
+    }
+}
+
+void initGyroData() {
+    for (int axis = 0; axis < 3; ++axis) {
+        for (int i = 0; i < FFT_WINDOW_SIZE; ++i) {
+            gyroData[axis][i] = 0;
+        }
+    }
+}
+
+static inline int fftFreqToBin(int freq) {
+    return ((FFT_WINDOW_SIZE / 2 - 1) * freq) / (fftMaxFreq);
+}
+
+void gyroDataAnalyseInit(uint32_t targetLooptimeUs) {
+    // initialise even if FEATURE_DYNAMIC_FILTER not set, since it may be set later
+    samplingFrequency = 1000000 / targetLooptimeUs;
+    fftSamplingScale = samplingFrequency / FFT_SAMPLING_RATE;
+    fftMaxFreq = FFT_SAMPLING_RATE / 2;
+    fftBinCount = fftFreqToBin(fftMaxFreq) + 1;
+    fftResolution = FFT_SAMPLING_RATE / FFT_WINDOW_SIZE;
+    arm_rfft_fast_init_f32(&fftInstance, FFT_WINDOW_SIZE);
+
+    initGyroData();
+    initHanning();
+
+    // recalculation of filters takes 4 calls per axis => each filter gets updated every 3 * 4 = 12 calls
+    // at 4khz gyro loop rate this means 4khz / 4 / 3 = 333Hz => update every 3ms
+    float looptime = targetLooptimeUs * 4 * 3;
+    for (int axis = 0; axis < 3; ++axis) {
+        fftResult[axis].centerFreq = 200; // any init value
+        biquadFilterInitLPF(&fftFreqFilter[axis], DYN_NOTCH_CHANGERATE, looptime);
+        biquadFilterInit(&fftGyroFilter[axis], FFT_BPF_HZ, 1000000 / FFT_SAMPLING_RATE, BIQUAD_Q, FILTER_BPF);
+    }
+}
 
-void gyroDataAnalyseInit(void)
-{
+// used in OSD
+const gyroFftData_t *gyroFftData(int axis) {
+    return &fftResult[axis];
 }
 
-void gyroDataAnalyse(const gyroDev_t *gyroDev, const gyro_t *gyro)
-{
-    UNUSED(gyroDev);
-    UNUSED(gyro);
+bool isDynamicFilterActive(void) {
+    return (IS_RC_MODE_ACTIVE(BOXDYNAMICFILTER) || feature(FEATURE_DYNAMIC_FILTER));
 }
 
-void gyroDataAnalyseUpdate(timeUs_t currentTimeUs)
-{
-    UNUSED(currentTimeUs);
+/*
+ * Collect gyro data, to be analysed in gyroDataAnalyseUpdate function
+ */
+void gyroDataAnalyse(const gyroDev_t *gyroDev, biquadFilter_t *notchFilterDyn) {
+    if (!isDynamicFilterActive()) {
+        return;
+    }
+
+    // if gyro sampling is > 1kHz, accumulate multiple samples
+    for (int axis = 0; axis < 3; ++axis) {
```

Comment:
Preincrement is fine.

Postincrement is way C programmers are used to write it, because there is no difference in C. 
In C++ preincrement is cheaper (with overload), so using it in both languages will result in net gain ...